### PR TITLE
JIT: arm64 osr funclet frame probing

### LINKED ISSUE #70263: Test failure: JIT\Regression\JitBlue\GitHub_21990\GitHub_21990\GitHub_21990.cmd
runtime-jit-experimental, `jit_osr_stress_random`

```
D:\h\w\A6B3094B\w\B97F09EC\e>type D:\h\w\A6B3094B\w\B97F09EC\u\SetStressModes_jitosr_stress_random.cmd 
set COMPlus_TieredCompilation=1
set COMPlus_TC_OnStackReplacement=1
set COMPlus_TC_QuickJitForLoops=1
set COMPlus_TC_OnStackReplacement_InitialCounter=1
set COMPlus_OSR_HitLimit=2
set COMPlus_JitRandomOnStackReplacement=15

D:\h\w\A6B3094B\w\B97F09EC\e>set PATH=D:\h\w\A6B3094B\p\dotnet-cli;C:\python3.9.1\Scripts\;C:\python3.9.1\;C:\python3.7.0\Scripts\;C:\python3.7.0\;D:\Windows\system32;D:\Windows;D:\Windows\System32\Wbem;D:\Windows\System32\WindowsPowerShell\v1.0\;D:\Windows\System32\OpenSSH\;D:\Users\runner\AppData\Local\Microsoft\WindowsApps 

D:\h\w\A6B3094B\w\B97F09EC\e>set DOTNET_ROOT=D:\h\w\A6B3094B\p\dotnet-cli 

D:\h\w\A6B3094B\w\B97F09EC\e>set DOTNET_CLI_TELEMETRY_OPTOUT=1 

D:\h\w\A6B3094B\w\B97F09EC\e>set DOTNET_CLI_HOME=D:\h\w\A6B3094B\w\B97F09EC\e\.dotnet 

D:\h\w\A6B3094B\w\B97F09EC\e>set NUGET_PACKAGES=D:\h\w\A6B3094B\w\B97F09EC\e\.nuget 

D:\h\w\A6B3094B\w\B97F09EC\e>dotnet D:\h\w\A6B3094B\p\xunit\xunit.console.dll JIT\Regression\JIT.Regression.XUnitWrapper.dll -parallel collections -nocolor -noshadow -xml testResults.xml -trait TestGroup=JIT.Regression.JitBlue 
Microsoft.DotNet.XUnitConsoleRunner v2.5.0 (64-bit .NET 7.0.0-preview.3.22175.4)
  Discovering: JIT.Regression.XUnitWrapper (method display = ClassAndMethod, method display options = None)
  Discovered:  JIT.Regression.XUnitWrapper (found 227 of 1601 test cases)
  Starting:    JIT.Regression.XUnitWrapper (parallel test collections = on, max threads = 8)
    JIT\Regression\JitBlue\GitHub_21990\GitHub_21990\GitHub_21990.cmd [FAIL]
      
      Return code:      1
      Raw output file:      D:\h\w\A6B3094B\w\B97F09EC\uploads\Reports\JIT.Regression\JitBlue\GitHub_21990\GitHub_21990\GitHub_21990.output.txt
      Raw output:
      BEGIN EXECUTION
       "D:\h\w\A6B3094B\p\corerun.exe" -p "System.Reflection.Metadata.MetadataUpdater.IsSupported=false"  GitHub_21990.dll 
      Expected: 100
      Actual: -1073741819
      END EXECUTION - FAILED
      FAILED
      Test Harness Exitcode is : 1
      To run the test:
      > set CORE_ROOT=D:\h\w\A6B3094B\p
      > D:\h\w\A6B3094B\w\B97F09EC\e\JIT\Regression\JitBlue\GitHub_21990\GitHub_21990\GitHub_21990.cmd
      Expected: True
      Actual:   False
      Stack Trace:
           at JIT_Regression._JitBlue_GitHub_21990_GitHub_21990_GitHub_21990_._JitBlue_GitHub_21990_GitHub_21990_GitHub_21990_cmd()
      Output:
        
        Return code:      1
        Raw output file:      D:\h\w\A6B3094B\w\B97F09EC\uploads\Reports\JIT.Regression\JitBlue\GitHub_21990\GitHub_21990\GitHub_21990.output.txt
        Raw output:
        BEGIN EXECUTION
         "D:\h\w\A6B3094B\p\corerun.exe" -p "System.Reflection.Metadata.MetadataUpdater.IsSupported=false"  GitHub_21990.dll 
        Expected: 100
        Actual: -1073741819
        END EXECUTION - FAILED
        FAILED
        Test Harness Exitcode is : 1
        To run the test:
        > set CORE_ROOT=D:\h\w\A6B3094B\p
        > D:\h\w\A6B3094B\w\B97F09EC\e\JIT\Regression\JitBlue\GitHub_21990\GitHub_21990\GitHub_21990.cmd
  Finished:    JIT.Regression.XUnitWrapper
=== TEST EXECUTION SUMMARY ===
   JIT.Regression.XUnitWrapper  Total: 227, Errors: 0, Failed: 1, Skipped: 0, Time: 23.349s
```

https://dev.azure.com/dnceng/public/_build/results?buildId=1807282&view=ms.vss-test-web.build-test-results-tab

### PR DESCRIPTION
**Pull request by AndyAyersMS**
We currently need to pad OSR funclet frames with the full size of the
Tier0 frame, and this padding sits above the register save area. For
large Tier0 frames we thus need to probe the stack in the OSR prolog
for type1 and type3 frames, before doing the initial SP adjustment and
save.

Longer term the plan is to revise arm64 OSR so we will not need to
do this padding; at that point we can remove this extra probing.

Fixes #70263.

### CODE REVIEW

**Comment by AndyAyersMS** [comment]
@BruceForstall PTAL
cc @dotnet/jit-contrib

File: src/coreclr/jit/codegenarm64.cpp [hunk 1]
```diff
@@ -1354,10 +1354,12 @@ void CodeGen::genFuncletProlog(BasicBlock* block)
 
     if (genFuncletInfo.fiFrameType == 1)
     {
-        // With OSR we may see large values for fiSpDelta1
-        // (we really need to probe the frame, sigh)
         if (compiler->opts.IsOSR())
         {
+            // With OSR we may see large values for fiSpDelta1.
+            // We repurpose genAllocLclFram to do the necessary probing.
```
Comment (jkotas) [inline src/coreclr/jit/codegenarm64.cpp:1360]:
```suggestion
            // We repurpose genAllocLclFrame to do the necessary probing.
```

File: src/coreclr/jit/codegenarm64.cpp [hunk 2]
```diff
@@ -1393,10 +1395,12 @@ void CodeGen::genFuncletProlog(BasicBlock* block)
     }
     else if (genFuncletInfo.fiFrameType == 3)
     {
-        // With OSR we may see large values for fiSpDelta1
-        // (we really need to probe the frame, sigh)
         if (compiler->opts.IsOSR())
         {
+            // With OSR we may see large values for fiSpDelta1
+            // We repurpose genAllocLclFram to do the necessary probing.
```
Comment (jkotas) [inline src/coreclr/jit/codegenarm64.cpp:1401]:
```suggestion
            // We repurpose genAllocLclFrame to do the necessary probing.
```

**Comment by BruceForstall** [comment]
Can you show what the funclet prolog from https://github.com/dotnet/runtime/issues/70263#issuecomment-1159214254 looks like with this change?

**Comment by BruceForstall** [comment]
What about the `genStackPointerAdjustment(genFuncletInfo.fiSpDelta1, REG_SCRATCH, nullptr, /* reportUnwindData */ true);` done for frame type 5? What range is `genFuncletInfo.fiSpDelta1` in that case? (can there be asserts added that indicate the maximum range?)

**Comment by jakobbotsch** [comment]
"Frame type 5" is already quite broken, see #66089.

**Comment by AndyAyersMS** [comment]
> Can you show what the funclet prolog from [#70263 (comment)](https://github.com/dotnet/runtime/issues/70263#issuecomment-1159214254) looks like with this change?

Old/new prologs

<table>
<tr>
<td valign="bottom">

```asm
mov     x9, #0x9ce0
sub     sp, sp, x9
stp     fp, lr, [sp]
stp     x19, x20, [sp,#32]
mov     x2, #0x9ce0
add     x3, fp, x2
str     x3, [sp,#24]
```

</td>
<td>

```asm
movn    x9, #0xfff
movn    x1, #0x9cdf
ldr     wzr, [sp, x9]
sub     x9, x9, #1, LSL #12
cmp     x1, x9
bls     pc-16 (-4 instructions)
mov     x9, #0x9ce0
sub     sp, sp, x9
stp     fp, lr, [sp]
stp     x19, x20, [sp,#32]
mov     x2, #0x9ce0
add     x3, fp, x2
str     x3, [sp,#24]
```

</td>
</tr>
</table>

**Comment by AndyAyersMS** [comment]
caught up in the guid change

**Comment by AndyAyersMS** [comment]
Libraries failure looks like a gc hole perhaps:
```
Assert failure(PID 5564 [0x000015bc], Thread: 708 [0x02c4]): m_alignpad == 0
```

**Comment by jkotas** [comment]
> Assert failure(PID 5564 [0x000015bc], Thread: 708 [0x02c4]): m_alignpad == 0

#70231

**Comment by BruceForstall** [comment]
You should rebase after https://github.com/dotnet/runtime/pull/70922, and I think you need the same change for fiFrameType=5.

**Comment by BruceForstall** [comment]
> You don't mind that I am repurposing `genAllocLclFrame` like this? It is (more or less) allocating the "tier0" frame equivalent in the funclet.

It seems ok, since `genAllocLclFrame` on arm64 only does frame probing (the name is a misnomer for arm64). So as long as it's probing the right amount, for OSR, it makes sense.

**Review by BruceForstall — APPROVED** [review]

**Comment by AndyAyersMS** [comment]
Failure looks like #69474

**Comment by AndyAyersMS** [comment]
Jit-experimental was clean other than an OSX timeout.

Other parts of this pipeline run the Github_21990 test case under OSR stress, and it's been consistently failing there on arm64 for a long time.

**Pull request merged**